runtime/compiler/control/CompilationThread.cpp

```diff
@@ -11637,6 +11637,11 @@ TR::CompilationInfo::storeAOTInSharedCache(
                 * to send two different buffers to store in cache and also as no one queries either code data / metadata for
                 * method, clean up the share classs cache API.
                 */
+               TR_ASSERT(comp, "The AOT compilation that succeeded must have a compilation object.");
+               J9JITDataCacheHeader *aotMethodHeader = (J9JITDataCacheHeader *)comp->getAotMethodDataStart();
```

Comment:
I see that the code in the master branch looks different, so I assume we merged master into jitaas and subsequently fixes were added to the compression code. In that case it's better to do another merge and take the fixes in rather than patching this code.

Reply:
I'm not sure I follow. This looks like a mistake was made during the last merge. 

This `storeAOTInSharedCache` routine seems to have been introduced in the JITServer branch here. We refactored this code out of `compilationEnd`: https://github.com/eclipse/openj9/pull/4788

In upstream we recently added this new patch of code into `compilationEnd`: https://github.com/eclipse/openj9/pull/5926/.

During the last merge, that patch of code was brought into `storeAOTInSharedCache`. However, the code uses variables that do not exist inside `storeAOTInSharedCache`. The change here resolves that problem by bringing down the definitions of the variables from `compilationEnd` into `storeAOTInSharedCache`.

So I don't think there were any fixes delivered in upstream as that code looks correct.

Note: We could alternatively pass these variables into `storeAOTInSharedCache` from `compilationEnd` if we don't want to define them again. Let me know if you prefer that.

Reply:
Actually I just found a test case where we hit the assert here because `aotMethodHeader` is null. Investigating.

Reply:
Dhruv's explanation is correct. I moved the compression code into `storeAOTInSharedCache` during the latest merge. But since compression is not enabled for x86, I could not test if I moved code correctly, and it turns out I didn't.

Reply:
I've disabled this compression code for JITServer: b2aff5daef

Reply:
Note: I disabled this because `aotMethodDataStart` is NULL at the time that we execute this code. In certain scenarios (AOT Compilation fails), we do set the `aotMethodDataStart` later on: https://github.com/eclipse/openj9/blob/855783cc343a7e92ea1e2ce47d4f58191193931d/runtime/compiler/control/JITaaSCompilationThread.cpp#L2953

@mpirvu : Is this expected? Or does something need to change here?

Reply:
The caller of `storeAOTInSharedCache` already performs checks on `aotMethodDataStart`:
```
               J9JITDataCacheHeader *aotMethodHeader      = (J9JITDataCacheHeader *)comp->getAotMethodDataStart();
               TR_ASSERT(aotMethodHeader, "The header must have been set");
               TR_AOTMethodHeader   *aotMethodHeaderEntry = (TR_AOTMethodHeader *)(aotMethodHeader + 1);
```
so the assert at the caller should fire first.
With or without compression the `aotMethodDataStart` should have been set by now, so if that doesn't happen we have a bug.

Reply:
Looking again at the code, when the client is invoking a remote AOT compilation, at the end of it it calls: `remoteCompilationEnd()` which calls `TR::CompilationInfo::storeAOTInSharedCache()`. At this point the client has not yet set `aotMethodDataStart`. This is done somewhat later and it's actually not used. We could move that `comp->setAotMethodDataStart(metaData);` up in the code and in that case we could allow AOT compression to happen.

Reply:
I would argue that inside `storeAOTInSharedCache()`, dataStart is the J9JITDataCacheHeader, so we can easily get to our `aotMethodHeaderEntry`

Reply:
@mpirvu I can move the code that sets the meta data above. 

Quick question though: We only set the `aotMethodDataStart` if `canRelocateMethod==false`. However there's a comment here that this behaviour is different from baseline and that we don't use the data start for anything:

https://github.com/eclipse/openj9/blob/ceebd6063b47a56d2da8612941f4fabf2e1de5a0/runtime/compiler/control/JITaaSCompilationThread.cpp#L2997-L3000

 I just want to make sure with you that this won't cause conflict inside `storeAOTInSharedCache()`.

Reply:
My preference would be to modify `storeAOTInSharedCache()` to infer aotMethodHeaderEntry from the `dataStart` parameter.

Reply:
I have made the changes here: fa2aacc3706a002f2ba096ff6694238597b90180